README.md
# flask-uploadfile
[![CodeFactor](https://www.codefactor.io/repository/github/guan840912/flask-uploadfile/badge/master)](https://www.codefactor.io/repository/github/guan840912/flask-uploadfile/overview/master)
[![codecov](https://codecov.io/gh/guan840912/flask-uploadfile/branch/master/graph/badge.svg)](https://codecov.io/gh/guan840912/flask-uploadfile)
![Publish Docker](https://github.com/guan840912/flask-uploadfile/workflows/Publish%20Docker/badge.svg)
[![License](https://img.shields.io/badge/License-Apache%202.0-yellowgreen.svg)](https://opensource.org/licenses/Apache-2.0)

### This for easy upload file via python flask . 
- enable login page 
- uploadfile 
- downloadfile
- docker image   
### you can run in docker .

### also can run in local 
```bash
$ git clone https://github.com/guan840912/flask-uploadfile.git

$ cd flask-uploadfile/

$ export URL='0.0.0.0'

$ python app.py

open your browser http://localhost:8080

# default login admin/admin
```

### Usage Docker  
```bash
docker run -d -p 8080:8080 guanyebo/flask-uploadfile
```

### Usage Docker-compose 
```bash
docker-compose up -d
```



### Use curl and wget  , uploadfile and downloadfile and deletefile
```bash
$ touch aa.tar

$ curl -X POST -F file=@"aa.tar" http://localhost:8080/uploadfile

$ wget http://localhost:8080/downloadfile/aa.tar

$ curl -X DELETE http://localhost:8080/deletefile/aa.tar
```


## Note for run testing need to do first export ENV
```bash
export PASSWORD=admin
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.7.7](https://github.com/guan840912/flask-uploadfile/compare/v1.7.6...v1.7.7) (2020-10-10)

### [1.7.6](https://github.com/guan840912/flask-uploadfile/compare/v1.7.4...v1.7.6) (2020-09-26)


### Bug Fixes

* **core:** add codecov ignore ([8402e21](https://github.com/guan840912/flask-uploadfile/commit/8402e21b3cd9e51f4dee1af11c2a25590598feec))
* **core:** codecov ([c8450f2](https://github.com/guan840912/flask-uploadfile/commit/c8450f22c6cf6d044bf018796a10fa1eaf6b6f37))
* **core:** return statement in finally block may swallow exception ([acb32b8](https://github.com/guan840912/flask-uploadfile/commit/acb32b84530ad39bfda4b47f8b5c35716117603c))
* **core:** return statement in finally block may swallow exception ([49e1646](https://github.com/guan840912/flask-uploadfile/commit/49e1646c907b4d04cb7320bb282f6e18d3585002))
* **core:** return statement in finally block may swallow exception return None ([c22d571](https://github.com/guan840912/flask-uploadfile/commit/c22d5713a3809558f6f77ce65ed8dcc62d7c6498))

### [1.7.8](https://github.com/guan840912/flask-uploadfile/compare/v1.7.4...v1.7.8) (2020-09-26)


### Bug Fixes

* **core:** add codecov ignore ([8402e21](https://github.com/guan840912/flask-uploadfile/commit/8402e21b3cd9e51f4dee1af11c2a25590598feec))
* **core:** codecov ([c8450f2](https://github.com/guan840912/flask-uploadfile/commit/c8450f22c6cf6d044bf018796a10fa1eaf6b6f37))
* **core:** return statement in finally block may swallow exception ([acb32b8](https://github.com/guan840912/flask-uploadfile/commit/acb32b84530ad39bfda4b47f8b5c35716117603c))
* **core:** return statement in finally block may swallow exception ([49e1646](https://github.com/guan840912/flask-uploadfile/commit/49e1646c907b4d04cb7320bb282f6e18d3585002))
* **core:** return statement in finally block may swallow exception return None ([c22d571](https://github.com/guan840912/flask-uploadfile/commit/c22d5713a3809558f6f77ce65ed8dcc62d7c6498))

### [1.7.4](https://github.com/guan840912/flask-uploadfile/compare/v1.7.3...v1.7.4) (2020-09-15)

### [1.7.3](https://github.com/guan840912/flask-uploadfile/compare/v1.7.2...v1.7.3) (2020-07-28)

### [1.7.2](https://github.com/guan840912/flask-uploadfile/compare/v1.7.1...v1.7.2) (2020-07-27)

### [1.7.1](https://github.com/guan840912/flask-uploadfile/compare/v1.7.0...v1.7.1) (2020-07-27)

## [1.7.0](https://github.com/guan840912/flask-uploadfile/compare/v1.6.2...v1.7.0) (2020-07-27)

### [1.6.2](https://github.com/guan840912/flask-uploadfile/compare/v1.6.1...v1.6.2) (2020-07-27)

### [1.6.1](https://github.com/guan840912/flask-uploadfile/compare/v1.6.0...v1.6.1) (2020-07-26)

## [1.6.0](https://github.com/guan840912/flask-uploadfile/compare/v1.5.0...v1.6.0) (2020-07-26)

## [1.5.0](https://github.com/guan840912/flask-uploadfile/compare/v1.4.0...v1.5.0) (2020-07-26)

## [1.4.0](https://github.com/guan840912/flask-uploadfile/compare/v1.3.0...v1.4.0) (2020-07-26)

## [1.3.0](https://github.com/guan840912/flask-uploadfile/compare/v1.2.0...v1.3.0) (2020-07-26)

## [1.2.0](https://github.com/guan840912/flask-uploadfile/compare/v1.1.0...v1.2.0) (2020-07-24)

## [1.1.0](https://github.com/guan840912/flask-uploadfile/compare/v1.0.0...v1.1.0) (2020-07-24)


test.sh
x=1
while [ $x -eq 1  ]
do
   curl http://localhost:8080/ping
   sleep 1
done


tests.py
from app import app , get_file_size , creation_date
from unittest.mock import patch
import unittest
import io 
import os 

ppwd=os.getenv('PASSWORD')
class FlaskTestCase(unittest.TestCase):
    # Ensure that flask was set up correctly
    def test_index(self):
        tester = app.test_client(self)
        response = tester.get('/login', content_type='html/text')
        self.assertEqual(response.status_code, 200)

    # Ensure Please login in /login page.
    def test_login_page_loads(self):
        tester = app.test_client(self)
        response = tester.get('/login', content_type='html/text')
        self.assertTrue( b'Please login' in response.data)

    # Ensure can login in /login page.
    def test_can_currect_login_page(self):
        tester = app.test_client(self)
        response = tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        self.assertIn( b'Upload new File' , response.data,msg=None) 
        self.assertIn( b'Show Local file Table' , response.data,msg=None) 
        self.assertIn( b'show' , response.data,msg=None) 

    # Ensure incurrect login in /login page.
    def test_can_incurrect_login_page(self):
        tester = app.test_client(self)
        response = tester.post(
            '/login',
            data=dict(username="admin", password="xxxx"), 
            follow_redirects=True
        )
        self.assertIn( b'Invalid Credentials. Please try again.' , response.data,msg=None) 

    # Ensure Please login in /welcome page.
    def test_welcome_page_loads(self):
        tester = app.test_client(self)
        response = tester.get('/welcome', content_type='application/json')
        self.assertTrue( b'Welcome to Flask!' in response.data)

    # Ensure Please login in /ping page.
    def test_ping_payloads(self):
        tester = app.test_client(self)
        response = tester.get('/ping', content_type='html/text')
        self.assertTrue( b'{"message":"pong!!!"}' in response.data)

    # Ensure logout can use.
    def test_can_secuess_logout_page(self):
        tester = app.test_client(self)
        tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        response = tester.post(
            '/logout',
            follow_redirects=True
        )
        self.assertIn( b'Welcome to Flask!' , response.data,msg=None)
    
    # Ensure Upload new File in / page.
    def test_home_page_loads(self):
        tester = app.test_client(self)
        tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        response = tester.get('/', content_type='html/text')
        self.assertTrue( b'Upload new File' in response.data)
        self.assertIn( b'Show Local file Table' , response.data,msg=None) 
        self.assertIn( b'show' , response.data,msg=None)

    # Ensure Upload new File in / page.
    def test_show_page_loads(self):
        tester = app.test_client(self)
        tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        response = tester.get('/show', content_type='html/text')
        self.assertIn( b'List File' , response.data,msg=None) 

    # Ensure Upload File not in allow file .
    def test_upload_not_allow_type_file(self):
        tester = app.test_client(self)
        tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        response = tester.post('/uploadfile', data=dict(
            file=(io.BytesIO(b'hi everyone'), 'test.md'),
        ))
        self.assertEqual(response.status_code, 302)
    
    # Ensure Upload file api can use .
    def test_upload_file(self):
        tester = app.test_client(self)
        tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        response = tester.post('/uploadfile', data=dict(
            file=(io.BytesIO(b'hi everyone'), 'test.txt'),
        ))
        self.assertIn( b'Date' , response.data,msg=None) 

    def test_no_select_file_upload_file(self):
        tester = app.test_client(self)
        from io import BytesIO
        tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        response = tester.post('/uploadfile',data=dict(file=(BytesIO(b''),''),follow_redirects=True))
        self.assertEqual(response.status_code, 302)

    def test_download_file(self):
        tester = app.test_client(self)
        tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        response = tester.get('/downloadfile/.gitkeep')
        self.assertEqual(response.status_code, 200)

    def test_show_buttom(self):
        tester = app.test_client(self)
        tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        response = tester.post('/',data=dict(show='show'))
        self.assertEqual(response.status_code, 302)

    def test_logout_buttom(self):
        tester = app.test_client(self)
        tester.post(
            '/login',
            data=dict(username="admin", password=ppwd), 
            follow_redirects=True
        )
        response = tester.post('/',data=dict(logout='logout'))
        self.assertEqual(response.status_code, 302)

    def test_zdelete_file(self):
        tester = app.test_client(self)
        response = tester.delete('/deletefile/test.txt')
        self.assertIn( b'deleted' , response.data,msg=None) 

    def test_nologin_get_show(self):
        tester = app.test_client(self)
        response = tester.get('/show')
        self.assertEqual(response.status_code, 302)

    
    def test_get_file_size_bytes(self):
        with patch('os.path.getsize', return_value=0):
            res = get_file_size("pwdfile")
        self.assertEqual(res,"0 bytes ")

    def test_get_file_size_kb(self):
        with patch('os.path.getsize', return_value=1024*2):
            res = get_file_size("pwdfile")
        self.assertEqual(res,"2.0 KB ")

    def test_get_file_size_mb(self):
        with patch('os.path.getsize', return_value=1024*1024*2):
            res = get_file_size("pwdfile")
        self.assertEqual(res,"2.0 MB ")

    def test_get_file_size_none(self):
        res = get_file_size("pwdfile")
        self.assertEqual(res,None)

    def test_os_gettime_windows(self):
        with patch('platform.system', return_value='Windows'):
            with patch('os.path.getctime', return_value='time'):
                res = creation_date("pwdfile")
        self.assertEqual(res,'time')

if __name__ == '__main__':
    unittest.main()

time-test.py
import os

def format_table():
    from flask_table import Table, Col
    pic_file_list = os.listdir( './pic' )
    class ItemTable(Table):
        name = Col('Name')
        download_link = Col('Download Link')
    
    # Get some objects
    class Item(object):
        def __init__(self, name, download_link):
            self.name = name
            self.download_link = download_link
    items = []
    for i in pic_file_list: 
        items.append(Item(str(i),'http://nexus-registry.cathayins-dev.com/downloadfile/'+ str(i)))
    table = ItemTable(items)
    
    return table.__html__()



try-py/date-time.py
from datetime import datetime, timezone, timedelta
import os
import platform

tz = timezone(timedelta(hours=+8))

def what_time_is():
    tz = timezone(timedelta(hours=+8))
    datetime.now(tz).isoformat()
    time = datetime.now(tz).isoformat(timespec="seconds")
    return str("This time is  " + time )

def creation_date(path_to_file):
    if platform.system() == 'Windows':
        return os.path.getctime(path_to_file)   
    stat = os.stat(path_to_file)
    try:
        return stat.st_birthtime
    except AttributeError:
        # We're probably on Linux. No easy way to get creation dates here,
        # so we'll settle for when its content was last modified.
        return stat.st_mtime

def get_file_size(path_to_file):
    try:
        size = os.path.getsize(path_to_file)
        if size <= 1024:
            return str(size) + " bytes "
        if size > 1024 and 1024*1024 >= size:
            return str(round(size/1024, 2)) + " KB "
        if size > 1024*1024 and 1024*1024*1024 >= size:
            return str(round(size/(1024*1024),2)) + " MB "
        return "None"
    except ZeroDivisionError:
        pass


print(datetime.fromtimestamp(creation_date('./pic/knative_logpng.png')).isoformat(timespec="seconds"))
print(what_time_is())
print(get_file_size('./pic/knative_logpng.png'))

try-py/sample-flask-table.py
'''A example for creating a simple table within a working Flask app.
Our table has just two columns, one of which shows the name and is a
link to the item's page. The other shows the description.'''
from flask_table import Table, Col, LinkCol
from flask import Flask
import os 
from flask import send_from_directory , jsonify , send_file ,render_template

def download_file(name):
    return '''<a href="./downloadfile/{name}" download="{name}">
              {name}</a>'''.format(name=name)

app = Flask(__name__)


def format_table():
    pic_file_list = os.listdir( '../pic' )
    class ItemTable(Table):
        name = Col('Name')
        download_link = Col('Download Link')
    
    # Get some objects
    class Item(object):
        def __init__(self, name, download_link):
            self.name = name
            self.download_link = download_link
    items = []
    for i in pic_file_list: 
        items.append(Item(str(i), download_file(str(i))))
    table = ItemTable(items,border="1")
    
    return (table.__html__())


@app.route('/')
def index():
    listfile=''
    pic_file_list = os.listdir( '../pic' )
    for i in pic_file_list: 
        listfile += '<br>'+download_file(i)
    return render_template('show.html', table=listfile)
    #return table


if __name__ == '__main__':
    app.run()

app.py
# -*- coding: utf-8 -*-
import os
from datetime import datetime, timezone, timedelta
from flask import Flask, render_template, redirect, url_for, request, session, flash, jsonify, send_from_directory
from werkzeug.utils import secure_filename
import json
import platform
from functools import wraps
import secrets
from flask_jwt_extended import JWTManager, jwt_required, create_access_token, get_jwt_identity
jwt = JWTManager()

cwd = os.getcwd()
URL=os.getenv('URL')
UPLOAD_FOLDER = cwd + '/pic/'
ALLOWED_EXTENSIONS = set(['txt', 'pdf', 'png', 'jpg', 'jpeg', 'gif', 'zip','tar'])
app = Flask(__name__)
jwt_secret_key = secrets.token_hex(16)
app.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER
app.config['JWT_SECRET_KEY'] = jwt_secret_key
jwt.init_app(app)
# add secret key , example output, secret_key = 000d88cd9d90036ebdd237eb6b0db000
secret_key = secrets.token_hex(16)
app.secret_key = secret_key

# LET USER LOGIN FIRST
def login_required(f):
    @wraps(f)
    def wrap(*args, **kwargs):
        if 'logged_in' in session:
            return f(*args, **kwargs)       
        flash('You need to login first.')
        return redirect(url_for('login'))
    return wrap

# get file create time .
def creation_date(path_to_file):
    if platform.system() == 'Windows':
        return os.path.getctime(path_to_file)   
    stat = os.stat(path_to_file)
    try:
        return stat.st_birthtime
    except AttributeError:
        # We're probably on Linux. No easy way to get creation dates here,
        # so we'll settle for when its content was last modified.
        return stat.st_mtime

# print(datetime.fromtimestamp(creation_date('./pic/.gitkeep')).isoformat(timespec="seconds"))

# what time is it ??
def what_time_is():
    tz = timezone(timedelta(hours=+8))
    datetime.now(tz).isoformat()
    time = datetime.now(tz).isoformat(timespec="seconds")
    return  time

# get file size
def get_file_size(path_to_file):
    try:
        size = os.path.getsize(path_to_file)
        if size <= 1024:
            result = str(size) + " bytes "
        if size > 1024 and 1024*1024 >= size:
            result = str(round(size/1024, 2)) + " KB "
        if size > 1024*1024 and 1024*1024*1024 >= size:
            result = str(round(size/(1024*1024),2)) + " MB "
    except:
        result = None
    return result

# 解析檔案名稱
def allowed_file(filename):
    return '.' in filename and \
           filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS

def download_file(name):
    file_create_time = datetime.fromtimestamp(creation_date(cwd + "/pic/" + name)).isoformat(timespec="seconds")
    file_size = get_file_size(cwd + "/pic/" + name)
    return '''<div style=inline>
    <p>{file_create_time}</p>
    <p>{file_size}</p>
    <a href="./downloadfile/{name}" download="{name}">{name}</a>
    </div>'''.format(name=name , file_create_time=file_create_time, file_size=file_size)

# login route
@app.route('/login', methods=['GET', 'POST'])
def login():
    error = None
    if request.method == 'POST':
        if request.form['username'] != 'admin' or request.form['password'] != 'admin':
            error = 'Invalid Credentials. Please try again.'
        else:
            session['logged_in'] = True
            flash('Your already login !!!')
            return redirect(url_for('index_page'))
            
    return render_template('login.html', error=error)

@app.route('/login_jwt', methods=['POST'])
def login_jwt():
    if not request.is_json:
        return jsonify({"msg": "Missing JSON in request"}), 400

    username = request.json.get('username', None)
    password = request.json.get('password', None)
    if not username:
        return jsonify({"msg": "Missing username parameter"}), 400
    if not password:
        return jsonify({"msg": "Missing password parameter"}), 400

    if username != 'admin' or password != 'admin':
        return jsonify({"msg": "Bad username or password"}), 401

    # Identity can be any data that is json serializable
    access_token = create_access_token(identity=username)
    return jsonify(access_token=access_token), 200

@app.route('/ping')
def test():
    return jsonify({"message": "pong!!!" })

# show table .
@app.route('/show', methods=['GET', 'POST'])
@login_required
def show():
    listfile='''<div style=inline><p> File Create Time </p>
    <p></p><p></p><p></p><p></p>
    <p> Size </p>
    <p></p><p></p><p></p><p></p>
    <p> Name </p></div>'''
    pic_file_list = os.listdir( cwd + '/pic' )
    for i in pic_file_list: 
        listfile += download_file(i)+'<br>'
    return render_template('show.html', table=listfile)
    
#Download file
@app.route('/downloadfile/<filename>', methods=['GET'])
@jwt_required
def return_file(filename):
    return send_from_directory(directory='pic', filename=filename, as_attachment=True)

#Delete file
@app.route('/deletefile/<filename>', methods=['DELETE'])
@jwt_required
def delete_file(filename):
    os.remove(cwd + '/pic/' + filename)
    return {"status": "deleted"}, 201
    
# Upload file
@app.route('/uploadfile', methods=['POST'])
@jwt_required
def upload_file():
    error = None
    try:
        file = request.files['file']
        # if user does not select file, browser also
        # submit an empty part without filename
        if file.filename == '':
            error = 'No selected file'
            raise RuntimeError
        if file and allowed_file(file.filename):
            filename = secure_filename(file.filename)
            # 儲存在 app 本地端
            file.save(os.path.join(app.config['UPLOAD_FOLDER'], filename))

            successres = str("Date " + what_time_is() + " 上傳 " + filename + " 成功")

            res = render_template('uploadPage.html', successres=successres)
        if not allowed_file(file.filename):
            error = 'Not Allow File type , please upload type in txt, pdf, png, jpg, jpeg, gif, zip,tar'
            raise RuntimeError
    except RuntimeError:
        flash(error)
        res = redirect(url_for('index_page'))
    return res

# 建立上傳網址
@app.route('/', methods=['GET', 'POST'])
@login_required
def index_page():
    if request.form.get('show') == "show":
        return redirect(url_for('show'))

    if request.form.get('logout') == "logout":
        return redirect(url_for('logout'))

    return render_template('index.html') 

@app.route('/welcome')
def welcome():
    return render_template('welcome.html')  # render a template

# define logout 
@app.route('/logout', methods=['GET', 'POST'])
@login_required
def logout():
    session.pop('logged_in', None)
    flash('You already logout !!!')
    return redirect(url_for('welcome'))

# start the server
if __name__ == '__main__':
    app.run(host=URL,port=8080)